Makefile

```diff
@@ -156,7 +156,10 @@ generate: protos
 
 protos: bin/protoc-gen-gogoctrd ## generate protobuf
 	@echo "$(WHALE) $@"
+	$(eval TMPDIR := $(shell mktemp -d))
+	@mv vendor/github.com/containerd/containerd/api $(TMPDIR)
```

Comment:
Perhaps this would warrant a comment why we're doing this.

I'm not exactly sure what this is doing; would `-mod=mod` (or perhaps one of the other options) also work for this? https://golang.org/ref/mod#build-commands

> `-mod=mod` tells the go command to ignore the vendor directory and to automatically update `go.mod`, for example, when an imported package is not provided by any known module.

Reply:
still tracking this down, looks like the 2 copies of api/ directory confuses the heck out of the protoc

Reply:
Here's what happens when i don't remove the stuff from vendor/

```
2021-07-13T18:27:58.5303797Z + bin/protoc-gen-gogoctrd
2021-07-13T18:28:08.8675685Z + protos
2021-07-13T18:28:09.1989451Z /home/runner/work/containerd/containerd/src/github.com/containerd/containerd/api/events/container.proto: Input is shadowed in the --proto_path by "/home/runner/work/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/containerd/containerd/api/events/container.proto".  Either use the latter file as your input or reorder the --proto_path so that the former file's location comes first.
2021-07-13T18:28:09.2003614Z 2021/07/13 18:28:09 protoc -I./protobuf:./api:/home/runner/work/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/gogo/protobuf:/home/runner/work/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/gogo/googleapis:/home/runner/work/containerd/containerd/src/github.com/containerd/containerd/vendor:/home/runner/work/containerd/containerd/src/github.com/gogo/protobuf:/home/runner/work/containerd/containerd/src/github.com/gogo/googleapis:/home/runner/work/containerd/containerd/src:/usr/local/include:/usr/include --include_imports --descriptor_set_out=/tmp/descriptors.pb-204475644 --gogoctrd_out=plugins=fieldpath,import_path=github.com/containerd/containerd/api/events,Mgogoproto/gogo.proto=github.com/gogo/protobuf/gogoproto,Mgoogle/protobuf/any.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/descriptor.proto=github.com/gogo/protobuf/protoc-gen-gogo/descriptor,Mgoogle/protobuf/duration.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/empty.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/field_mask.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/timestamp.proto=github.com/gogo/protobuf/types,Mgoogle/rpc/status.proto=github.com/gogo/googleapis/google/rpc:/home/runner/work/containerd/containerd/src /home/runner/work/containerd/containerd/src/github.com/containerd/containerd/api/events/container.proto /home/runner/work/containerd/containerd/src/github.com/containerd/containerd/api/events/content.proto /home/runner/work/containerd/containerd/src/github.com/containerd/containerd/api/events/image.proto /home/runner/work/containerd/containerd/src/github.com/containerd/containerd/api/events/namespace.proto /home/runner/work/containerd/containerd/src/github.com/containerd/containerd/api/events/snapshot.proto /home/runner/work/containerd/containerd/src/github.com/containerd/containerd/api/events/task.proto
2021-07-13T18:28:09.2011586Z Makefile:158: recipe for target 'protos' failed
```

Reply:
Oh! Does it work it you run it from within the api directory? If you're outside of the api directory, you're in the "containerd" module, so the api directory is no longer part of containerd (so "doesn't exist", similar to how `go test ./...` will ignore the api directory)

Reply:
interesting, will check

Reply:
@thaJeztah moved the Protobuild.toml as well but does not work (added a cd to api/ directory in the Makefile target) it's a black box :(

```
[dims@dims-a02 17:41] ~/go/src/github.com/containerd/containerd ⟩ make protos
🇩 bin/protoc-gen-gogoctrd
🇩 protos
2021/07/13 17:41:59 bad output from command: github.com/containerd/containerd
github.com/containerd/containerd/archive
github.com/containerd/containerd/archive/compression
github.com/containerd/containerd/archive/tartest
github.com/containerd/containerd/cio
github.com/containerd/containerd/cmd/containerd
github.com/containerd/containerd/cmd/containerd/command
github.com/containerd/containerd/cmd/containerd-shim
github.com/containerd/containerd/cmd/containerd-stress
github.com/containerd/containerd/cmd/ctr
github.com/containerd/containerd/cmd/ctr/app
github.com/containerd/containerd/cmd/ctr/commands
github.com/containerd/containerd/cmd/ctr/commands/containers
github.com/containerd/containerd/cmd/ctr/commands/content
github.com/containerd/containerd/cmd/ctr/commands/events
github.com/containerd/containerd/cmd/ctr/commands/images
github.com/containerd/containerd/cmd/ctr/commands/install
github.com/containerd/containerd/cmd/ctr/commands/leases
github.com/containerd/containerd/cmd/ctr/commands/namespaces
github.com/containerd/containerd/cmd/ctr/commands/oci
github.com/containerd/containerd/cmd/ctr/commands/plugins
github.com/containerd/containerd/cmd/ctr/commands/pprof
github.com/containerd/containerd/cmd/ctr/commands/run
github.com/containerd/containerd/cmd/ctr/commands/shim
github.com/containerd/containerd/cmd/ctr/commands/snapshots
github.com/containerd/containerd/cmd/ctr/commands/tasks
github.com/containerd/containerd/cmd/ctr/commands/version
github.com/containerd/containerd/cmd/gen-manpages
github.com/containerd/containerd/cmd/protoc-gen-gogoctrd
github.com/containerd/containerd/containers
github.com/containerd/containerd/content
github.com/containerd/containerd/content/local
github.com/containerd/containerd/content/proxy
github.com/containerd/containerd/content/testsuite
github.com/containerd/containerd/contrib/apparmor
github.com/containerd/containerd/contrib/nvidia
github.com/containerd/containerd/contrib/seccomp
github.com/containerd/containerd/contrib/snapshotservice
github.com/containerd/containerd/defaults
github.com/containerd/containerd/diff
github.com/containerd/containerd/diff/apply
github.com/containerd/containerd/diff/walking
github.com/containerd/containerd/diff/walking/plugin
github.com/containerd/containerd/errdefs
github.com/containerd/containerd/events
github.com/containerd/containerd/events/exchange
github.com/containerd/containerd/filters
github.com/containerd/containerd/gc
github.com/containerd/containerd/gc/scheduler
github.com/containerd/containerd/identifiers
github.com/containerd/containerd/images
github.com/containerd/containerd/images/archive
github.com/containerd/containerd/images/converter
github.com/containerd/containerd/images/converter/uncompress
github.com/containerd/containerd/labels
github.com/containerd/containerd/leases
github.com/containerd/containerd/leases/proxy
github.com/containerd/containerd/log
github.com/containerd/containerd/log/logtest
github.com/containerd/containerd/metadata
github.com/containerd/containerd/metadata/boltutil
github.com/containerd/containerd/mount
github.com/containerd/containerd/namespaces
github.com/containerd/containerd/oci
github.com/containerd/containerd/pkg/apparmor
github.com/containerd/containerd/pkg/atomic
github.com/containerd/containerd/pkg/cri
github.com/containerd/containerd/pkg/cri/annotations
github.com/containerd/containerd/pkg/cri/config
github.com/containerd/containerd/pkg/cri/constants
github.com/containerd/containerd/pkg/cri/io
github.com/containerd/containerd/pkg/cri/opts
github.com/containerd/containerd/pkg/cri/server
github.com/containerd/containerd/pkg/cri/server/bandwidth
github.com/containerd/containerd/pkg/cri/server/testing
github.com/containerd/containerd/pkg/cri/store
github.com/containerd/containerd/pkg/cri/store/container
github.com/containerd/containerd/pkg/cri/store/image
github.com/containerd/containerd/pkg/cri/store/label
github.com/containerd/containerd/pkg/cri/store/sandbox
github.com/containerd/containerd/pkg/cri/store/snapshot
github.com/containerd/containerd/pkg/cri/store/truncindex
github.com/containerd/containerd/pkg/cri/streaming
github.com/containerd/containerd/pkg/cri/streaming/portforward
github.com/containerd/containerd/pkg/cri/streaming/remotecommand
github.com/containerd/containerd/pkg/cri/util
github.com/containerd/containerd/pkg/dialer
github.com/containerd/containerd/pkg/ioutil
github.com/containerd/containerd/pkg/netns
github.com/containerd/containerd/pkg/os
github.com/containerd/containerd/pkg/os/testing
github.com/containerd/containerd/pkg/process
github.com/containerd/containerd/pkg/progress
github.com/containerd/containerd/pkg/registrar
github.com/containerd/containerd/pkg/runtimeoptions/v1
github.com/containerd/containerd/pkg/seccomp
github.com/containerd/containerd/pkg/seed
github.com/containerd/containerd/pkg/seutil
github.com/containerd/containerd/pkg/stdio
github.com/containerd/containerd/pkg/testutil
github.com/containerd/containerd/pkg/timeout
github.com/containerd/containerd/pkg/ttrpcutil
github.com/containerd/containerd/pkg/userns
github.com/containerd/containerd/platforms
github.com/containerd/containerd/plugin
github.com/containerd/containerd/protobuf/plugin
github.com/containerd/containerd/protobuf/plugin/fieldpath
github.com/containerd/containerd/reference
github.com/containerd/containerd/reference/docker
github.com/containerd/containerd/remotes
github.com/containerd/containerd/remotes/docker
github.com/containerd/containerd/remotes/docker/auth
github.com/containerd/containerd/remotes/docker/config
github.com/containerd/containerd/remotes/docker/schema1
github.com/containerd/containerd/remotes/errors
github.com/containerd/containerd/rootfs
github.com/containerd/containerd/runtime
github.com/containerd/containerd/runtime/linux/runctypes
github.com/containerd/containerd/runtime/restart
github.com/containerd/containerd/runtime/restart/monitor
github.com/containerd/containerd/runtime/v1
github.com/containerd/containerd/runtime/v1/shim
github.com/containerd/containerd/runtime/v1/shim/client
github.com/containerd/containerd/runtime/v1/shim/v1
github.com/containerd/containerd/runtime/v2
github.com/containerd/containerd/runtime/v2/logging
github.com/containerd/containerd/runtime/v2/runc/options
github.com/containerd/containerd/runtime/v2/shim
github.com/containerd/containerd/runtime/v2/task
github.com/containerd/containerd/services
github.com/containerd/containerd/services/containers
github.com/containerd/containerd/services/content
github.com/containerd/containerd/services/content/contentserver
github.com/containerd/containerd/services/diff
github.com/containerd/containerd/services/events
github.com/containerd/containerd/services/healthcheck
github.com/containerd/containerd/services/images
github.com/containerd/containerd/services/introspection
github.com/containerd/containerd/services/leases
github.com/containerd/containerd/services/namespaces
github.com/containerd/containerd/services/opt
github.com/containerd/containerd/services/server
github.com/containerd/containerd/services/server/config
github.com/containerd/containerd/services/snapshots
github.com/containerd/containerd/services/tasks
github.com/containerd/containerd/services/version
github.com/containerd/containerd/snapshots
github.com/containerd/containerd/snapshots/native
github.com/containerd/containerd/snapshots/native/plugin
github.com/containerd/containerd/snapshots/proxy
github.com/containerd/containerd/snapshots/storage
github.com/containerd/containerd/snapshots/testsuite
github.com/containerd/containerd/sys
github.com/containerd/containerd/sys/reaper
github.com/containerd/containerd/version

make: *** [protos] Error 1
[dims@dims-a02 17:41] ~/go/src/github.com/containerd/containerd ⟩ make protos
🇩 bin/protoc-gen-gogoctrd
🇩 protos
/Users/dims/go/src/github.com/containerd/containerd/api
2021/07/13 17:42:23 bad output from command: github.com/containerd/containerd
github.com/containerd/containerd/archive
github.com/containerd/containerd/archive/compression
github.com/containerd/containerd/archive/tartest
github.com/containerd/containerd/cio
github.com/containerd/containerd/cmd/containerd
github.com/containerd/containerd/cmd/containerd/command
github.com/containerd/containerd/cmd/containerd-shim
github.com/containerd/containerd/cmd/containerd-stress
github.com/containerd/containerd/cmd/ctr
github.com/containerd/containerd/cmd/ctr/app
github.com/containerd/containerd/cmd/ctr/commands
github.com/containerd/containerd/cmd/ctr/commands/containers
github.com/containerd/containerd/cmd/ctr/commands/content
github.com/containerd/containerd/cmd/ctr/commands/events
github.com/containerd/containerd/cmd/ctr/commands/images
github.com/containerd/containerd/cmd/ctr/commands/install
github.com/containerd/containerd/cmd/ctr/commands/leases
github.com/containerd/containerd/cmd/ctr/commands/namespaces
github.com/containerd/containerd/cmd/ctr/commands/oci
github.com/containerd/containerd/cmd/ctr/commands/plugins
github.com/containerd/containerd/cmd/ctr/commands/pprof
github.com/containerd/containerd/cmd/ctr/commands/run
github.com/containerd/containerd/cmd/ctr/commands/shim
github.com/containerd/containerd/cmd/ctr/commands/snapshots
github.com/containerd/containerd/cmd/ctr/commands/tasks
github.com/containerd/containerd/cmd/ctr/commands/version
github.com/containerd/containerd/cmd/gen-manpages
github.com/containerd/containerd/cmd/protoc-gen-gogoctrd
github.com/containerd/containerd/containers
github.com/containerd/containerd/content
github.com/containerd/containerd/content/local
github.com/containerd/containerd/content/proxy
github.com/containerd/containerd/content/testsuite
github.com/containerd/containerd/contrib/apparmor
github.com/containerd/containerd/contrib/nvidia
github.com/containerd/containerd/contrib/seccomp
github.com/containerd/containerd/contrib/snapshotservice
github.com/containerd/containerd/defaults
github.com/containerd/containerd/diff
github.com/containerd/containerd/diff/apply
github.com/containerd/containerd/diff/walking
github.com/containerd/containerd/diff/walking/plugin
github.com/containerd/containerd/errdefs
github.com/containerd/containerd/events
github.com/containerd/containerd/events/exchange
github.com/containerd/containerd/filters
github.com/containerd/containerd/gc
github.com/containerd/containerd/gc/scheduler
github.com/containerd/containerd/identifiers
github.com/containerd/containerd/images
github.com/containerd/containerd/images/archive
github.com/containerd/containerd/images/converter
github.com/containerd/containerd/images/converter/uncompress
github.com/containerd/containerd/labels
github.com/containerd/containerd/leases
github.com/containerd/containerd/leases/proxy
github.com/containerd/containerd/log
github.com/containerd/containerd/log/logtest
github.com/containerd/containerd/metadata
github.com/containerd/containerd/metadata/boltutil
github.com/containerd/containerd/mount
github.com/containerd/containerd/namespaces
github.com/containerd/containerd/oci
github.com/containerd/containerd/pkg/apparmor
github.com/containerd/containerd/pkg/atomic
github.com/containerd/containerd/pkg/cri
github.com/containerd/containerd/pkg/cri/annotations
github.com/containerd/containerd/pkg/cri/config
github.com/containerd/containerd/pkg/cri/constants
github.com/containerd/containerd/pkg/cri/io
github.com/containerd/containerd/pkg/cri/opts
github.com/containerd/containerd/pkg/cri/server
github.com/containerd/containerd/pkg/cri/server/bandwidth
github.com/containerd/containerd/pkg/cri/server/testing
github.com/containerd/containerd/pkg/cri/store
github.com/containerd/containerd/pkg/cri/store/container
github.com/containerd/containerd/pkg/cri/store/image
github.com/containerd/containerd/pkg/cri/store/label
github.com/containerd/containerd/pkg/cri/store/sandbox
github.com/containerd/containerd/pkg/cri/store/snapshot
github.com/containerd/containerd/pkg/cri/store/truncindex
github.com/containerd/containerd/pkg/cri/streaming
github.com/containerd/containerd/pkg/cri/streaming/portforward
github.com/containerd/containerd/pkg/cri/streaming/remotecommand
github.com/containerd/containerd/pkg/cri/util
github.com/containerd/containerd/pkg/dialer
github.com/containerd/containerd/pkg/ioutil
github.com/containerd/containerd/pkg/netns
github.com/containerd/containerd/pkg/os
github.com/containerd/containerd/pkg/os/testing
github.com/containerd/containerd/pkg/process
github.com/containerd/containerd/pkg/progress
github.com/containerd/containerd/pkg/registrar
github.com/containerd/containerd/pkg/runtimeoptions/v1
github.com/containerd/containerd/pkg/seccomp
github.com/containerd/containerd/pkg/seed
github.com/containerd/containerd/pkg/seutil
github.com/containerd/containerd/pkg/stdio
github.com/containerd/containerd/pkg/testutil
github.com/containerd/containerd/pkg/timeout
github.com/containerd/containerd/pkg/ttrpcutil
github.com/containerd/containerd/pkg/userns
github.com/containerd/containerd/platforms
github.com/containerd/containerd/plugin
github.com/containerd/containerd/protobuf/plugin
github.com/containerd/containerd/protobuf/plugin/fieldpath
github.com/containerd/containerd/reference
github.com/containerd/containerd/reference/docker
github.com/containerd/containerd/remotes
github.com/containerd/containerd/remotes/docker
github.com/containerd/containerd/remotes/docker/auth
github.com/containerd/containerd/remotes/docker/config
github.com/containerd/containerd/remotes/docker/schema1
github.com/containerd/containerd/remotes/errors
github.com/containerd/containerd/rootfs
github.com/containerd/containerd/runtime
github.com/containerd/containerd/runtime/linux/runctypes
github.com/containerd/containerd/runtime/restart
github.com/containerd/containerd/runtime/restart/monitor
github.com/containerd/containerd/runtime/v1
github.com/containerd/containerd/runtime/v1/shim
github.com/containerd/containerd/runtime/v1/shim/client
github.com/containerd/containerd/runtime/v1/shim/v1
github.com/containerd/containerd/runtime/v2
github.com/containerd/containerd/runtime/v2/logging
github.com/containerd/containerd/runtime/v2/runc/options
github.com/containerd/containerd/runtime/v2/shim
github.com/containerd/containerd/runtime/v2/task
github.com/containerd/containerd/services
github.com/containerd/containerd/services/containers
github.com/containerd/containerd/services/content
github.com/containerd/containerd/services/content/contentserver
github.com/containerd/containerd/services/diff
github.com/containerd/containerd/services/events
github.com/containerd/containerd/services/healthcheck
github.com/containerd/containerd/services/images
github.com/containerd/containerd/services/introspection
github.com/containerd/containerd/services/leases
github.com/containerd/containerd/services/namespaces
github.com/containerd/containerd/services/opt
github.com/containerd/containerd/services/server
github.com/containerd/containerd/services/server/config
github.com/containerd/containerd/services/snapshots
github.com/containerd/containerd/services/tasks
github.com/containerd/containerd/services/version
github.com/containerd/containerd/snapshots
github.com/containerd/containerd/snapshots/native
github.com/containerd/containerd/snapshots/native/plugin
github.com/containerd/containerd/snapshots/proxy
github.com/containerd/containerd/snapshots/storage
github.com/containerd/containerd/snapshots/testsuite
github.com/containerd/containerd/sys
github.com/containerd/containerd/sys/reaper
github.com/containerd/containerd/version

make: *** [protos] Error 1
```

Reply:
I wonder if it needs to be split (as the `Protobuild.toml` contains both things for the "containerd" module and for the "api" module 🤔.


I also noticed that the API depends on containerd, so because of that, there's a circular dependency (api module depends on containerd module, depends on api module); https://github.com/containerd/containerd/blob/c1b01eabc07f35b949650db27c87deafbe978f7d/api/events/container.proto#L23

Did some quick attempts to see if I could get it to work, but it's ugly (moved the makefile into the API directory, because it generates "packages" for which to generate protos, but doing so, it would consider them to be part of the "containerd" module (and using vending from that), not the containerd/api module.

https://github.com/containerd/containerd/compare/main...thaJeztah:wip_try_move_things?expand=1

Reply:
going back to the `mv` the `vendor/` directory out of the way here. things may take longer to make suitable changes to `protobuild`